Feature/secret sharing

kserapea opened the pull request:
@choct155 I believe that the Polynomial class is "ready ;)" to merge to main. Please review! Thanks!

choct155 commented:
@kserapea Here is a comment on the code review

choct155 commented on .gitignore:
```diff
@@ -3,6 +3,10 @@ __pycache__/
 
+# Intellij stuff
+.idea/
+*/.DS_Store
```
I hope this `.DS_Store` restriction works moving forward, but it probably won't for stuff that's already there.

choct155 commented on src/python/main.py:
```diff
@@ -0,0 +1,11 @@
+from IPython.display import display, Image
+
+def run() -> None:
+    ...
```
I like the use of the ellipsis placeholder!

kserapea replied:
I learn from the best 🥇

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+
+    def __str__(self) -> str:
+        """returns a string representation of the polynomial"""
+        coeff_list: List[str] = [display(c, i) for i, c in enumerate(self.coefficients)]
```
`display()` appears to be a standalone function, but I think that it is only related to the display of instances of `Polynomial`. So, I am not saying that you should require an instance of `Polynomial` on which to call it, but I do think that it is organizationally linked to the class `Polynomial`. In that case, it's probably better to define it as a [static method](https://pythonbasics.org/static-method/).  For example...

```python
class Polynomial:
    
    def __init__(self, ...):
        ...

    @staticmethod
    def display(coefficient: int, index: int) -> str:
        ...
```

kserapea replied:
I added as a staticmethod to the class (and also discovered that I needed to call it using `Polynomial.display` within the function. in order for it to work)

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+    def __str__(self) -> str:
+        """returns a string representation of the polynomial"""
+        coeff_list: List[str] = [display(c, i) for i, c in enumerate(self.coefficients)]
+        filter_coeff_list = list(filter(lambda x: x != "", coeff_list))
```
Ultimately, what you are doing is concatenating together (in line 14) a bunch of substrings. Cool. When talking about strings under concatenation, the empty string (i.e. `""`) is the inert value. What do I mean by that? When thinking about integers under addition, the inert value is 0. Which means, if I add 0 to any integer, I get that integer back. When thinking about integers under multiplication, the inert value is 1. The empty string, in the context of string concatenation, is inert. When I concatenate an empty string to `s: str` I get `s` back. 

Consequently, it feels like line 13 doesn't do anything at the end of the day. Maybe remove it?

kserapea replied:
If I didn't filter out the empty string, I was ending up with a blank space even when joining, so this is why I went and filtered out the empty string (see screenshot below of what would happen).

![Screen Shot 2022-01-03 at 8 52 39 AM](https://user-images.githubusercontent.com/58795090/147938546-cf747766-23e1-4028-a344-0734e6eba65f.png)

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+        """returns a string representation of the polynomial"""
+        coeff_list: List[str] = [display(c, i) for i, c in enumerate(self.coefficients)]
+        filter_coeff_list = list(filter(lambda x: x != "", coeff_list))
+        return re.sub(r"\+ -", "- ", " + ".join(filter_coeff_list))
```
What is the regex sub operation solving for?

kserapea replied:
This solves for a negative coefficient - removing the inserted "+" during the join and ensuring it is "-" (see screenshot below for the -2 coefficient)

![Screen Shot 2022-01-03 at 8 56 33 AM](https://user-images.githubusercontent.com/58795090/147938979-fe29ddce-85a3-4d2a-9837-bc9c7fd8159c.png)

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+
+    def __add__(self, other: "Polynomial") -> "Polynomial":
+        """returns the sum of two polynomials (like degrees are added together)"""
+        return [sum(n) for n in it.zip_longest(self.coefficients, other.coefficients, fillvalue = 0)]
```
What is the type of the returned value supposed to be? The signature suggests it should be of type `Polynomial`, but it returns a value of type `List[int]`.

kserapea replied:
I went fixed this type, thanks!

choct155 replied:
Does that mean you changed the type to `List[int]`?  If so, wouldn't you want `__add__()` to be an `(A, A) -> A` operation?

kserapea replied:
ok maybe this aligns with the question I just posed below for the similar case - are we saying I should in fact just wrap the output as a Polynomial? I think that makes sense, though I am not sure the pros/cons. But I think you're leading me.

choct155 replied:
Yep, just like below, I think you want a Polynomial output

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+        """returns the product of two polynomials"""
+        len_self: int = len(self.coefficients)
+        len_other: int = len(other.coefficients)
+        blank_coeff_list: list(int) = [0] * (len_self + len_other - 1)
```
Do we know that the resulting `Polynomial` will always include `len_self + len_other - 1` terms after simplification? Feels like a nice opportunity for a proof.

kserapea replied:
So we had actually already proved a few things when we did the [chapter 2.1 exercise](https://github.com/kserapea/pim/blob/feature/secret_sharing/Chapter%202%20Exercises.ipynb).

- If 𝑓 is a degree 2 polynomial and 𝑔 is a degree 1 polynomial then their product 𝑓⋅𝑔 is a degree 3 polynomial
- If 𝑓 is a degree 𝑛 polynomial and 𝑔 is a degree 𝑚 polynomial then their product 𝑓⋅𝑔 is a degree 𝑛+𝑚 polynomial.
- If 𝑓 or 𝑔 are a zero polynomial, their product 𝑓⋅𝑔 is a degree 𝑚 polynomial (convention that a zero polynomial has a degree of 0).

I will think through the code to ensure it actually covers those cases instead of using my unproven (and subsequently disproved method).

kserapea replied:
I realize now that what my 
`len_self` + `len_other` - 1 
was accounting for was that inherently, when you multiply x^1 * x^1, you end up with x^2, (which is of length 1); therefore I was automatically discarding the "0x" that would be present in a coefficient list. So, my thinking is not wrong for polynomials in general, but if we are measuring the length of a coefficient list and including 0's as placeholders to account for degrees, than the length aspect that way does not work. (you probably saw that right away, but I was thinking from the maths side and not from the data structures side...)

choct155 replied:
I don't think I saw it clearly right away, just had a nagging feeling. Thanks for confirming!

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+        #Multiplies two polynomials term by term
+        for i in range(len_self):
+            for j in range(len_other):
+                blank_coeff_list[i + j] += self.coefficients[i] * other.coefficients[j]
```
Suppose we have two instances of `Polynomial` `p1` and `p2`:

```python 
p1 : Polynomial = Polynomial([1])
p2 : Polynomial = Polynomial([1])
```

Then `p1 * p2` should be something that represents `x**2`, which would be `Polynomial([0, 1])`. It seems like the output value `blank_coeff_list` would have been `[1]` with this implementation. Is that correct?

Relatedly, this function returns a value of type `List[int]`. Just like above, is that desired? Or should it be of type `Polynomial`?

kserapea replied:
I'm fixing the first half of your question, but for the second half I'm slightly confused. Because the way we enter a Polynomial initially is a List[int]. If I did want to return a `Polynomial` is that simply me wrapping the output in `Polynomial`? Why would or wouldn't I want to do that?
`g: Polynomial = Polynomial([-8, 17, 0, 5])`

choct155 replied:
It is just wrapping like you have there. You would want to do that because the behavior on an instance of `Polynomial` is different than the behavior of the list. Also, just conceptually, if you multiply two polynomials together, shouldn't you get a polynomial back?

kserapea replied:
Ok I get that - thanks! (also I should refresh the page because I don't see your comments and then just keep doing Sarah things haha)

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+
+    def evaluate(self, value: int) -> int:
+        """evaluates a polynomial at a given value"""
+        calculate_list: list[str] = [calculate(c, i) for i, c in enumerate(self.coefficients)]
```
First, just like with `display()`, I suspect that `calculate()` is a function that only relates to instances of `Polynomial`. As such, I would recommend making it static.

Second, `calculate_list` seems like it should be of type `List[Tuple[int, int]]` based upon the function signature of `calculate()`. Is that correct?

kserapea replied:
I added calculate as a staticmethod to the class.

I also fixed the output list to be a List[Polyterm] ....more on that later.

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -0,0 +1,89 @@
+
+    return output
+
+def calculate(coefficient: int, index: int) -> (int, int):
```
You are returning a `Tuple[int, int]`, but the concept is a pairing of coefficient and order. That is, you are returning a single polynomial term. One thing, you could do for clarity (which I often do) is create a very simple [`namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple). 

```python
PolyTerm = namedtuple("PolyTerm", ["coefficient", "order"])
```

So this would imply, that body of your function would look something like the following:

```python
def calculate(coefficient: int, index: int) -> PolyTerm:
    if coefficient == 0:
        return PolyTerm(0, 0)
    else:
        return PolyTerm(coefficient, index)
```

Once you have have an instance of `PolyTerm`, you can reference things directly by name:

```python
pt: PolyTerm = PolyTerm(1, 2)
assert pt.coefficient == 1
assert pt.order == 2
```

This is not material from a mechanical perspective, but might make it easier to read the code.

kserapea replied:
I adjusted the calculate and evaluate functions to use a `PolyTerm` _namedtuple_; however I have not figured out yet how to actually make it realize I have defined `PolyTerm`. (I get the error of `name 'PolyTerm' is not defined` when I try to run the test code. <- This error was fixed by putting PolyTerm in quotes in the function signature. Just let me know if I should be housing the namedtuple somewhere else?

choct155 replied:
I am guessing that if you move the definition of `PolyTerm` in front of `Polynomial` (where you use it), you probably don't have to put it in quotes.

choct155 reviewed (commented):
@kserapea This looks great, but I had a few questions. Can you go through and address the comments that I left in the code?

kserapea commented:
I believe I answered/fixed the code you left comments on. There are two functions at the end `interpolate` and `single_term` which I transferred over from page 23 of PIM. However, I get an attribute error in line 89 of single term when it tries to use my sum function `AttributeError: 'list' object has no attribute 'coefficients'` and I haven't figured that out yet.

choct155 commented:
Hmmm, for some reason I feel like I already answered this last comment. Maybe I never sent or something. In any event, sounds like you are passing a value of type `List[int]` when you are expecting to have a value of type `Polynomial` at your disposal.

kserapea commented:
So 2 things:
1. His functions all use `floats`, and so there is a misalignment with me using `ints` for the polynomial - but I don't love the idea of switching all mine to `floats` but maybe we should just do that?
2. The line `terms: Polynomial = Polynomial([single_term(points, i) for i in range(0, len(points))])` is wrong. I'm the one who added the Polynomial wrapper, but what is inside there is actually a list of Polynomial - which I think is weird. But I think that is what I need to try to fix.

kserapea commented:
I fixed the attribute error, and modified the PIM functions, so it now works to interpolate one single set of points. However, when trying to input two sets of points I get a `IndexError: list index out of range` pointing to my `__mul__` function.

kserapea commented:
I wrote a function to find the degree of the polynomial (might be a little hacky?) - still getting a list index out of range error for the PIM interpolate function for 2 points - but will look at that again.

choct155 commented:
> I wrote a function to find the degree of the polynomial (might be a little hacky?) - still getting a list index out of range error for the PIM interpolate function for 2 points - but will look at that again.

Did this ever get resolved?

kserapea commented:
Unfortunately, no, I haven't figured it out yet.

kserapea commented:
Ok - it works now, I can show you what I changed, but it doesn't make sense to me. I changed it on a whim just to see what would happen (e.g. the debugger wasn't helping at all and I could really use some tips as to why it doesn't do what I think it should).

choct155 commented on src/python/pim/polynomial_representation.py:
```diff
@@ -28,12 +28,12 @@ def __mul__(self, other: "Polynomial") -> "Polynomial":
         """
         if Polynomial.find_degree(self) == -1:
-            degree_self: int = 0
+            degree_self: int = 1
```
I take it this is the change that fixed the `IndexError`? I presume all the tests pass now.
